//! The rurminer errors

use failure::{Backtrace, Context, Fail};
use std::fmt::{self, Debug, Display};

use std::io;

pub struct Error {
    inner: Context<ErrorKind>,
}

#[derive(Clone, Eq, PartialEq, Debug, Fail)]
pub enum ErrorKind {
    /// Standard input/output error.
    #[fail(display = "IO error: {}", _0)]
    Io(String),

    /// General error used for more specific input/output error.
    #[fail(display = "General error: {}", _0)]
    General(String),

    /// Error generated by backend for selected target.
    #[fail(display = "Backend error: {}", _0)]
    Backend(String),
}

/// Implement Fail trait instead of use Derive to get more control over custom type.
/// The main advantage is customization of Context type which allows conversion of
/// any error types to this custom error with general error kind by calling context
/// method on any result type.
impl Fail for Error {
    fn cause(&self) -> Option<&Fail> {
        self.inner.cause()
    }

    fn backtrace(&self) -> Option<&Backtrace> {
        self.inner.backtrace()
    }
}

impl Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        Display::fmt(&self.inner, f)
    }
}

impl Debug for Error {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        Debug::fmt(&self.inner, f)
    }
}

impl Error {
    pub fn kind(&self) -> ErrorKind {
        self.inner.get_context().clone()
    }
}

impl From<ErrorKind> for Error {
    fn from(kind: ErrorKind) -> Self {
        Self {
            inner: Context::new(kind),
        }
    }
}

impl From<Context<ErrorKind>> for Error {
    fn from(inner: Context<ErrorKind>) -> Self {
        Self { inner }
    }
}

impl From<io::Error> for Error {
    fn from(e: io::Error) -> Self {
        let msg = e.to_string();
        Self {
            inner: e.context(ErrorKind::Io(msg)),
        }
    }
}

impl From<Context<&str>> for Error {
    fn from(context: Context<&str>) -> Self {
        Self {
            inner: context.map(|info| ErrorKind::General(info.to_string())),
        }
    }
}

impl From<Context<String>> for Error {
    fn from(context: Context<String>) -> Self {
        Self {
            inner: context.map(|info| ErrorKind::General(info)),
        }
    }
}

#[cfg(feature = "backend_selected")]
pub mod backend {
    use super::*;
    use crate::hal;

    impl From<hal::Error> for Error {
        fn from(hal_error: hal::Error) -> Self {
            let msg = hal_error.to_string();
            Self {
                inner: hal_error.context(ErrorKind::Backend(msg)),
            }
        }
    }

    impl From<hal::ErrorKind> for Error {
        fn from(kind: hal::ErrorKind) -> Self {
            Self {
                inner: Context::new(ErrorKind::Backend(kind.to_string())),
            }
        }
    }

    impl From<Context<hal::ErrorKind>> for Error {
        fn from(context: Context<hal::ErrorKind>) -> Self {
            Self {
                inner: context.map(|info| ErrorKind::Backend(info.to_string())),
            }
        }
    }
}

/// A specialized `Result` type bound to [`Error`].
pub type Result<T> = std::result::Result<T, Error>;
